### PR TITLE
Monster icons

### PR DESCRIPTION
**Pull request by ArturKnopik**
Co-Authored-By: @Zbizu 
```LUA
creature:hasIcon(iconId)
creature:getIconValue(iconId)
creature:setIconValue(iconId, value)
creature:removeIcon(iconId)

monster:hasMonsterIcon(iconId)
monster:getMonsterIconValue(iconId)
monster:setMonsterIconValue(iconId, value)
monster:removeMonsterIcon(iconId)
```

image

### Pull Request Prelude
- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### CODE REVIEW

File: src/luascript.cpp [hunk 1]
```diff
@@ -8615,6 +8655,71 @@ int LuaScriptInterface::luaCreatureGetZone(lua_State* L)
 	return 1;
 }
 
+int LuaScriptInterface::luaCreatureHasIcon(lua_State* L)
```
Comment (MillhioreBT) [inline src/luascript.cpp:8658]: I think you should follow the format of the other methods:
```cpp
if (!usedata) {
lua_pushnil();
return 1;
}

-- Code
```

File: src/creature.cpp [hunk 2]
```diff
@@ -79,6 +79,17 @@ bool Creature::canSeeCreature(const Creature* creature) const
 	return true;
 }
 
+void Creature::refreshCreatureIcons() const
+{
+	SpectatorVec spectators;
+	g_game.map.getSpectators(spectators, this->getPosition(), true, true);
+	for (Creature* spectator : spectators) {
+		if (auto player = spectator->getPlayer()) {
```
Comment (MillhioreBT) [inline src/creature.cpp:87]: I think that here should also be verified in this way:
```cpp
assert(dynamic_cast<Player*>(spectator) != nullptr);
static_cast<Player*>(spectator)->sendUpdateCreatureIcons(this);
```

Comment (Zbizu) [reply src/creature.cpp:87]: Spectator existing and being player is already verified in `getSpectators` (`onlyPlayers` param). Would be nice if PR author could at least mention where he got that code from.

Comment (ArturKnopik) [reply src/creature.cpp:87]: Idk from where i took this code, it's possible it is your code :)
i have it for a while... :D

Comment (ArturKnopik) [reply src/creature.cpp:87]: image

**Comment by EPuncker** [comment]
nice!

**Comment by MillhioreBT** [comment]
I would like the names of the methods to be a little better...
Example:
`Creature.getIcons`
`Monster.getSpecialIcons`

Since the redundancy of names is something strange.
`Creature.getCreatureIcons`
`Monster.getMonsterIcons`

Unfortunately, they cannot be used generically as it would make it more difficult to manage. However, I think this should have a more appropriate name. maybe?

<details>

<summary>Other Things...</summary>

Wouldn't it be better to have two methods to handle or read the maps? this way they would be handled directly without creating methods that are never actually used on the sources side, they will only be used in lua

<details>

<summary>creature.h</summary>

```h
using CreatureIconHashMap = std::unordered_map<CreatureIcon_t, uint16_t>;

	// creature icons
	CreatureIconHashMap& getIcons() { return creatureIcons; }
	const CreatureIconHashMap& getIcons() const { return creatureIcons; }
	void updateIcons() const;

protected:
	CreatureIconHashMap creatureIcons;
```

</details>

<details>

<summary>monster.h</summary>

```h
using MonsterIconHashMap = std::unordered_map<MonsterIcon_t, uint16_t>;

	// monster icons
	MonsterIconHashMap& getSpecialIcons() { return monsterIcons; }
	const MonsterIconHashMap& getSpecialIcons() const { return monsterIcons; }

private:
	MonsterIconHashMap monsterIcons;
```

</details>

<details>

<summary>luascript - creature</summary>

```cpp
int LuaScriptInterface::luaCreatureHasIcon(lua_State* L)
{
	// creature:hasIcon(iconId)
	const Creature* creature = getUserdata<const Creature>(L, 1);
	if (creature) {
		pushBoolean(L, creature->getIcons().contains(getInteger<CreatureIcon_t>(L, 2)));
	} else {
		lua_pushnil(L);
	}
	return 1;
}

int LuaScriptInterface::luaCreatureSetIcon(lua_State* L)
{
	// creature:setIcon(iconId, value)
	Creature* creature = getUserdata<Creature>(L, 1);
	if (!creature) {
		lua_pushnil(L);
		return 1;
	}

	auto iconId = getInteger<CreatureIcon_t>(L, 2);
	if (iconId >= CREATURE_ICON_LAST) {
		reportErrorFunc(L, "Invalid Creature Icon Id");
		pushBoolean(L, false);
		return 1;
	}

	creature->getIcons()[iconId] = getInteger<uint16_t>(L, 3);
	creature->updateIcons();
	pushBoolean(L, true);
	return 1;
}

int LuaScriptInterface::luaCreatureGetIcon(lua_State* L)
{
	// creature:getIcon(iconId)
	Creature* creature = getUserdata<Creature>(L, 1);
	if (!creature) {
		lua_pushnil(L);
		return 1;
	}

	auto iconId = getInteger<CreatureIcon_t>(L, 2);
	const auto& icons = creature->getIcons();
	if (icons.contains(iconId)) {
		lua_pushinteger(L, icons.at(iconId));
	} else {
		lua_pushinteger(L, 0);
	}
	return 1;
}

int LuaScriptInterface::luaCreatureRemoveIcon(lua_State* L)
{
	// creature:removeIcon(iconId)
	Creature* creature = getUserdata<Creature>(L, 1);
	if (!creature) {
		lua_pushnil(L);
		return 1;
	}

	auto iconId = getInteger<CreatureIcon_t>(L, 2);
	auto& icons = creature->getIcons();
	if (icons.contains(iconId)) {
		icons.erase(iconId);
		creature->updateIcons();
		pushBoolean(L, true);
	} else {
		pushBoolean(L, false);
	}
	return 1;
}
```

</details>

<details>

<summary>luascript - monster</summary>

```cpp
int LuaScriptInterface::luaMonsterHasIcon(lua_State* L)
{
	// monster:hasSpecialIcon(iconId)
	const Monster* monster = getUserdata<const Monster>(L, 1);
	if (monster) {
		pushBoolean(L, monster->getSpecialIcons().contains(getInteger<MonsterIcon_t>(L, 2)));
	} else {
		lua_pushnil(L);
	}
	return 1;
}

int LuaScriptInterface::luaMonsterSetIcon(lua_State* L)
{
	// monster:setSpecialIcon(iconId, value)
	Monster* monster = getUserdata<Monster>(L, 1);
	if (!monster) {
		lua_pushnil(L);
		return 1;
	}

	auto iconId = getInteger<MonsterIcon_t>(L, 2);
	if (iconId >= MONSTER_ICON_LAST) {
		reportErrorFunc(L, "Invalid Monster Icon Id");
		pushBoolean(L, false);
		return 1;
	}

	monster->getSpecialIcons()[iconId] = getInteger<uint16_t>(L, 3);
	monster->updateIcons();
	pushBoolean(L, true);
	return 1;
}

int LuaScriptInterface::luaMonsterGetIcon(lua_State* L)
{
	// monster:getSpecialIcon(iconId)
	Monster* monster = getUserdata<Monster>(L, 1);
	if (!monster) {
		lua_pushnil(L);
		return 1;
	}

	auto iconId = getInteger<MonsterIcon_t>(L, 2);
	const auto& icons = monster->getSpecialIcons();
	if (icons.contains(iconId)) {
		lua_pushinteger(L, icons.at(iconId));
	} else {
		lua_pushinteger(L, 0);
	}
	return 1;
}

int LuaScriptInterface::luaMonsterRemoveIcon(lua_State* L)
{
	// monster:removeSpecialIcon(iconId)
	Monster* monster = getUserdata<Monster>(L, 1);
	if (!monster) {
		lua_pushnil(L);
		return 1;
	}

	auto iconId = getInteger<MonsterIcon_t>(L, 2);
	auto& icons = monster->getSpecialIcons();
	if (icons.contains(iconId)) {
		icons.erase(iconId);
		monster->updateIcons();
		pushBoolean(L, true);
	} else {
		pushBoolean(L, false);
	}
	return 1;
}
```

</details>

<details>

<summary>protocolgame.cpp</summary>

```cpp
void ProtocolGame::AddCreatureIcons(NetworkMessage& msg, const Creature* creature)
{
	const auto& creatureIcons = creature->getIcons();
	if (const Monster* monster = creature->getMonster()) {
		const auto& monsterIcons = monster->getSpecialIcons();
		msg.addByte(creatureIcons.size() + monsterIcons.size());
		for (const auto& [iconId, level] : monsterIcons) {
			msg.addByte(iconId);
			msg.addByte(1);
			msg.add<uint16_t>(level);
		}
	} else {
		msg.addByte(creatureIcons.size());
	}

	for (const auto& [iconId, level] : creatureIcons) {
		msg.addByte(iconId);
		msg.addByte(0);
		msg.add<uint16_t>(level);
	}
}
```

</details>

</details>

**Pull request closed without merging**